drivers: spi_nrfx_spim: Fixes related to 32 Mbps speed on nRF5340

anangl opened the pull request:
According to the nRF5340 PS, SPIM4 only supports 32 Mbps when the application core is running at 128 MHz. Drive configuration H0H1 must also be used for its pins for this speed.

Fixes #37221.

anangl committed:
drivers: spi_nrfx_spim: Fix limiting of SPIM frequency on nRF5340

According to the nRF5340 PS, SPIM4 only supports 32 Mbps when
the application core is running at 128 MHz. This patch adds
the corresponding check.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: spi_nrfx_spim: Change the way the nrfx driver is configured

According to the nRF5340 PS, for 32 Mbps high-speed SPI using SPIM4,
drive configuration H0H1 must be used. The underlying nrfx_spim driver
does it properly in its initialization function, so change the shim to
(re)initialize the driver when the SPI configuration is to be changed
(only then the speed to use is known), to avoid the need of duplicating
the corresponding code in the shim itself.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

lairdjm commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -143,14 +146,52 @@ static int configure(const struct device *dev,
+	/* For 32 Mbps speed on nRF5340, drive configuration H0H1 must be used
+	 * (see the Pin assignments section in the nRF5340 PS).
+	 */
+	nrf_gpio_pin_drive_t drive = (spim_frequency == NRF_SPIM_FREQ_32M)
```
What if someone wants to use H1H0 on slower speeds?

JordanYates replied:
I do have this open #34854, but it may need to be updated to do the init at this location instead of at driver init.

anangl replied:
> What if someone wants to use H1H0 on slower speeds?

It is not supported now by the driver, so I actually don't know what I could do about it in this PR. Do you have any suggestion?

JordanYates commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -143,14 +146,52 @@ static int configure(const struct device *dev,
+	/* For 32 Mbps speed on nRF5340, drive configuration H0H1 must be used
+	 * (see the Pin assignments section in the nRF5340 PS).
+	 */
+	nrf_gpio_pin_drive_t drive = (spim_frequency == NRF_SPIM_FREQ_32M)
```
Just realised that the HAL drivers already perform this configuration in `nrfx_spim_init`

anangl replied:
I'm sorry that it took me so long to get back to this one. I reworked the PR to take advantage of the pin configuration performed by the nrfx driver. Please take a look once again.

anangl committed:
drivers: spi_nrfx_spi: Change the way the nrfx driver is configured

Apply the same changes as the previous commit made in the spi_nrfx_spim
shim, to keep these two shims aligned.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Rebased.

carlescufi commented:
@gmarull can you approve if happy please?

gmarull commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -346,13 +339,14 @@ static int spim_nrfx_pm_control(const struct device *dev,
-		ret = init_spim(dev);
-		/* Force reconfiguration before next transfer */
-		data->ctx.config = NULL;
+		/* No action needed at this point, nrfx_spim_init() will be
+		 * called at configuration before the next transfer.
+		 */
```
PM hooks are no longer doing what is expected, ie, put the device into operational state when resume is called. Why is this change required?

anangl replied:
Previously, when resume was called, the device was brought to the state it was right after its initialization - it still needed to be reconfigured before the next transfer:
https://github.com/zephyrproject-rtos/zephyr/blob/7ccc1a41bc4c27c73bb03f01ed6b05491ba5872d/drivers/spi/spi_nrfx_spim.c#L336-L338
(at the device initialization, the nrfx driver was initialized with a default configuration, which was then altered accordingly when a transfer was requested). Now the difference is that the initialization of the device does not include the initialization of the nrfx driver - this is done when the device is configured right before a transfer is performed (only then we know the frequency, SPI mode, etc. that should be used), so there is no point in calling `nrfx_spim_init()` here and then, right before the next transfer, calling `nrfx_spim_uninit()` in order to be able to call `nrfx_spim_init()` again.
Alternatively, I would need to store the currently used nrfx driver configuration to be able to use it here in the call to `nrfx_spim_init()`. Then it would be not needed to reconfigure the device before the next transfer, provided it would be requested with the same configuration. But I'm not sure if it is worth spending more RAM just to achieve this possibility.

gmarull commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -286,13 +282,14 @@ static int spi_nrfx_pm_control(const struct device *dev,
-		data->ctx.config = NULL;
+		/* No action needed at this point, nrfx_spi_init() will be
+		 * called at configuration before the next transfer.
+		 */
```
same as previous comment

gmarull reviewed (commented):
I only have one comment regarding PM.

carlescufi commented:
@JordanYates please take another look
